bin/varnishd/cache/cache_obj.c

```diff
@@ -220,17 +222,33 @@ void
 ObjExtend(struct worker *wrk, struct objcore *oc, ssize_t l, int final)
 {
 	const struct obj_methods *om = obj_getmethods(oc);
+	AN(om->objextend);
 
 	CHECK_OBJ_NOTNULL(wrk, WORKER_MAGIC);
 	CHECK_OBJ_NOTNULL(oc->boc, BOC_MAGIC);
 	(void)final;
 	assert(l >= 0);
 
 	Lck_Lock(&oc->boc->mtx);
-	AN(om->objextend);
 	if (l > 0) {
+		/* Pass readahead applies when only sending to one client */
+		if (oc->flags & OC_F_PRIVATE) {
+			/* delivered_so_far will be re-read due to memory clobber
+			   from the CondWait opaque function call */
+			while (!(oc->flags & OC_F_ABANDON) &&
+				oc->boc->pass_prefetch > 0 &&
+				oc->boc->fetched_so_far > oc->boc->delivered_so_far + oc->boc->pass_prefetch) {
+				/* Wait 1ms for client to catch up */
+				(void)Lck_CondWait(&oc->boc->cond, &oc->boc->mtx, VTIM_real() + 1e-3);
```

Comment:
Can you please comment on the reasons for the timeout?
In my mind, a perfect solution would ensure that we signal the CV under all possible circumstances and not use any timeout.
We might still decide on the compromise of a timeout to avoid mishaps, but I fear relevant overhead for 1000 CondWaits a seconds.

Reply:
I made the timeout 30ms now